GH-107465: Add `pathlib.Path.from_uri()` classmethod. (#107640)
python/cpython: Doc/library/pathlib.rst

   .. versionadded:: 3.5

.. classmethod:: Path.from_uri(uri)

   Return a new path object from parsing a 'file' URI conforming to
   :rfc:`8089`. For example::

       >>> p = Path.from_uri('file:///etc/hosts')
       PosixPath('/etc/hosts')

   On Windows, DOS device and UNC paths may be parsed from URIs::

       >>> p = Path.from_uri('file:///c:/windows')
       WindowsPath('c:/windows')
       >>> p = Path.from_uri('file://server/share')
       WindowsPath('//server/share')

   Several variant forms are supported::

       >>> p = Path.from_uri('file:////server/share')
       WindowsPath('//server/share')
       >>> p = Path.from_uri('file://///server/share')
       WindowsPath('//server/share')
       >>> p = Path.from_uri('file:c:/windows')
       WindowsPath('c:/windows')
       >>> p = Path.from_uri('file:/c|/windows')

>        WindowsPath('c:/windows')

Reviewer: I thought ``|`` was a banned character on Windows!

Author: [According to wikipedia](https://en.wikipedia.org/wiki/File_URI_scheme#Windows_2):

> On Microsoft Windows systems, the normal colon (:) after a device letter has sometimes been replaced by a vertical bar (|) in file URLs. This reflected the original URL syntax, which made the colon a reserved character in a path part.

It's covered in [RFC 8089 appendix E.2.2](https://datatracker.ietf.org/doc/html/rfc8089#appendix-E.2.2).

   URIs with no slash after the scheme (and no drive letter) are parsed as
   relative paths::

       >>> p = Path.from_uri('file:foo/bar')

>        WindowsPath('foo/bar')

Reviewer: Wikipedia (an authoritative reference...) [suggests](https://en.wikipedia.org/wiki/File_URI_scheme#How_many_slashes?) that URIs with the ``file`` scheme must always contain at least one slash:

> A valid file URI must therefore begin with either `file:/path` (no hostname), `file:///path` (empty hostname), or `file://hostname/path`.

Supposing that this is true, do we want to support non-standards-compliant behaviour?

Author: It's arguable. All the file URI standards are post-hoc, and sometimes file URIs are supported alongside [relative URLs](https://datatracker.ietf.org/doc/html/rfc1808), but in that case the `file:` scheme prefix is usually forbidden. E.g. `man 7 uri` says:

> Note that if you really mean to say "start from the current location," don't specify the scheme at all; use a relative address like `../test.txt`, which has the side-effect of being scheme-independent.

[`urllib.request.url2pathname()`](https://docs.python.org/3/library/urllib.request.html#urllib.request.url2pathname) supports relative paths and doesn't require the `file:` prefix, but it's very old and somewhat hard to use correctly.

A universal_pathlib user requested support for relative URIs with a `file:` prefix [here](https://github.com/fsspec/universal_pathlib/commit/662c2943deff680423e048f62326cfdb8405ef09#r123906140).

We could perhaps add a *strict* argument that rejects URIs unless they begin `file:/`?

Reviewer: > We could perhaps add a _strict_ argument that rejects URIs unless they begin `file:/`?

I'd support this, defaulting to True.

Author: I've made the implementation stricter. I'll look at adding a *non*-strict mode later, if there's demand.

>    Users may wish to test the result with :meth:`~PurePath.is_absolute` and
>    reject relative paths, as these are not portable across processes with
>    different working directories.

Reviewer: I found this sentence a little hard to follow
```suggestion
   Users may wish to test the result of :meth:`!from_uri` with 
   :meth:`~PurePath.is_absolute`, as relative paths are not portable
   across processes with different working directories.
```
